### PR TITLE
data file override support with precedence

### LINKED ISSUE #891: enable loading resources from XDG_DATA_HOME before hitting package dir
Enable loading probe resources (e.g. custom prompts) from XDG_DATA_HOME first, package dir second

Should be for 
* most hits to `resources` dir
* `payloads`

### PR DESCRIPTION
**Pull request by jmartin-tech**
Fix #891 

* Provides a custom path accessor for project `data` files to allow user's to override project provided datasets using `$XDG_DATA_DIR/garak/data` to mirror `garak/data` from the package.
* Separates non-code data or dataset files from source code
* Downloaded datasets are considered `cached files` and do not change location or behavior in the revision
* Removes argument parser no longer supported by `gcg`

Consumers of the `garak.data.path` object can traverse the data directory as a standard `Path` object and will be returned existing files based on the order precedence from the supported locations on disk.

A possible drawback to the current implementation exists regarding directory interaction, if a folder exists at more than one path location the higher precedence path will be returned, when performing `glob` or iteration actions with such an instance the only files in the selected path on disk will be evaluated.  This is currently considered an acceptable limitation.

## Verification

List the steps needed to make sure this thing works

- [ ] Run the tests and ensure they pass `python -m pytest tests/`
- [ ] `garak -m nim -p donotanswer.MisinformationHarms -g 1 -n meta/llama3-70b-instruct --parallel_attempts 5`
- [ ] create a limited data set for `donotanswer.MisinformationHarms`
```
mkdir -p ~/.local/share/garak/data/donotanswer
head -n 10 garak/data/donotanswer/misinformation_harms.txt > ~/.local/share/garak/data/donotanswer/misinformation_harms.txt
```
- [ ] `garak -m nim -p donotanswer.MisinformationHarms -g 1 -n meta/llama3-70b-instruct --parallel_attempts 5`
- [ ] **Verify** run results:
  - [ ] the first run reported `155` attempts
  - [ ] the second run report `10` attempts (count depends `-n` count passed to `head`)

Note: #870 provides new `data` files that will need to migrate to this pattern depending on PR landing order.

### CODE REVIEW

**Comment by jmartin-tech** [comment]
Looks like there is some work to do with support for this on [python 3.10](https://stackoverflow.com/questions/61689391/error-with-simple-subclassing-of-pathlib-path-no-flavour-attribute) apparently!

**Comment by leondz** [comment]
> A possible drawback to the current implementation exists regarding directory interaction, if a folder exists at more than one path location the higher precedence path will be returned, when performing glob or iteration actions with such an instance the only files in the selected path on disk will be evaluated. This is currently considered an acceptable limitation.

Flagging a tension here with payload enumeration. Currently payload enum globs but uses its own directory traversal code, rather than `garak.data`. The result is that this PR doesn't break payload enum, but does leave us with two bits of code using similar-but-not-identical logic.

File: garak/data/__init__.py
Comment (leondz) [inline garak/data/__init__.py:46]: Why not `FileNotFoundError`?

File: garak/data/__init__.py
Comment (leondz) [inline garak/data/__init__.py:57]: Why not `FileNotFoundError`?

File: garak/probes/topic.py
Comment (leondz) [inline garak/probes/topic.py:95]: What's our plan for cache dir structure?

File: tests/test_data.py
Comment (leondz) [inline tests/test_data.py:65]: What expectation is being tested here?

File: tools/termscrape.py
Comment (leondz) [inline tools/termscrape.py:25]: Since this is writing, perhaps it should be sent to cache and not data. Actually, perhaps it's worth retiring this script from the repo.

**Review by leondz — COMMENTED** [review]
A few questions, looks good so far. Will be some extra work involved in aligning the payloads PR to the proposals here. The general separation is for non-code that can also be overridden should move into data; the rest remains in resources.

File: garak/data/__init__.py
Comment (erickgalinkin) [reply garak/data/__init__.py:46]: Maybe it makes sense to add a garak-specific FNFE to `garak.exceptions`? `GarakPathNotFound` or something?

File: garak/detectors/riskywords.py
Comment (erickgalinkin) [inline garak/detectors/riskywords.py:15]: Probably worth looking at as part of a larger refactor -- do we want all these datasets in our git? Or should we have them in some other place _e.g._ HuggingFace hub, and the `garak.data` module can also manage downloading these files?

File: garak/resources/common.py
Comment (erickgalinkin) [inline garak/resources/common.py:55]: This sounds right -- the file should be downloaded to `data_path / "advbench" / "harmful_behaviors.csv"`. We can probably do away with anything below. Just use `advbench_base_path` -> `advbench_path` and use that everywhere. Once you have it, we should be referencing it in a single location.

File: garak/data/__init__.py
Comment (jmartin-tech) [reply garak/data/__init__.py:46]: For this error, the actual cause here is more akin to a `ValueError` as the raise is due to a relative path traversal  that is not allowed even if it does exists.

I used a `garak` specific exceptions as consumers of these objects `know` their file exists. To me this means an exception from this method indicates something else `caused` the file to be `removed`, which is different than `not found` IMO.

File: garak/data/__init__.py
Comment (jmartin-tech) [reply garak/data/__init__.py:57]: I used a `garak` specific exceptions as consumers of these objects `know` their file exists. To me this means an exception from this method indicates something else `caused` the file to be `removed`, which is different than `not found` IMO.

File: garak/resources/tap/tap_main.py
Comment (erickgalinkin) [inline garak/resources/tap/tap_main.py:38]: Since `probes.tap` references the `prompts_location` as `data_path / "tap" / "tap_jailbreaks.txt"`, do we want this to reflect that as well? The idea is to keep your TAP results such that you can use them (or share them with us!) in subsequent probes. Though perhaps that's a larger issue.

File: garak/probes/visual_jailbreak.py
Comment (jmartin-tech) [inline garak/probes/visual_jailbreak.py:52]: Here is a case where I think the cached file is probe specific and pathing reflects the intent.

File: garak/resources/common.py
Comment (jmartin-tech) [inline garak/resources/common.py:53]: This provides an example for a case where the tooling would allow override of a resource not shipped with `garak` by placing a file in the `data` path using a known location.

File: garak/probes/topic.py
Comment (jmartin-tech) [reply garak/probes/topic.py:95]: There is an example [here](https://github.com/leondz/garak/pull/916/files#diff-d0e6a61922e5a549cc4c628aabf0b38c7c8c7c707a83a4f367ca511cd46e609bR56-R61) where `data` prefix is used already for a `shared` file to maintain consistency, maybe this should stay `_config.transient.cache_dir / "data" / "wn"` and other locations that use `cache_dir` to store in a `tool` namespace should place that namespace under `_config.transient.cache_dir / "data"`.

File: tests/test_data.py
Comment (jmartin-tech) [reply tests/test_data.py:65]: This test proves out the order or precedence.

The test verifies a file will be found from the `package_dir` when it does not exist in the then creates a file that will override that location in the `data_dir` and verifies a new request for that filename will return the `data_dir` file.

I could expand the test to read and compare file contexts if that will make the intent more clear.

File: tools/termscrape.py
Comment (jmartin-tech) [reply tools/termscrape.py:25]: Glad you noticed something in this file, I needed to flag this file is moved to `tools` and need to validate that `tools` will not be picked up in the `pypi` packaging.

To answer the original question, this tool seemed targeted as a maintainer script that could be used to update the source `data` files.  Since the file should already exist in the repo this external tool `writing` to the file would not violate the `LocalDataPath` expectations that files accessed during `garak` runtime are treated as read only.

**Comment by jmartin-tech** [comment]
> Flagging a tension here with payload enumeration. Currently payload enum globs but uses its own directory traversal code, rather than garak.data. The result is that this PR doesn't break payload enum, but does leave us with two bits of code using similar-but-not-identical logic.

Depending on land order I am thinking this can be a use case to improve `glob` support on the custom `Path` objects to return a consolidated list of files based on searching all `ORDERED_SEARCH_PATHS`. This would then remove the `similar-but-not-identical logic` by supporting the need quickly as a next iteration.

File: garak/detectors/riskywords.py
Comment (jmartin-tech) [reply garak/detectors/riskywords.py:15]: Have discussed in the past that we may want tooling that will download all datasets to create an `offline` deployment capability.

I suspect there could be some expansion on `garak.data` for handling access to known datasets. Treating HF as the specific location that tooling can register data as available from might be a good direction.

File: garak/probes/topic.py
Comment (jmartin-tech) [inline garak/probes/topic.py:96]: Interesting item here, the `parent` must exist before download however if the `wn` folder already exists the raised when attempting access below will be a different exception type and fails to recover.

This was `masked` previously by the fact that `_config.transient.cache_dir / "resources"` is created in early initialization for the `plugin_cache.json` file.

File: garak/probes/topic.py
Comment (leondz) [reply garak/probes/topic.py:95]: so far in `cache` I see:

* plugin cache
* wn
* visual_jailbreak data

It seems advantageous to be able to share artefacts across plugins, so I disprefer the idea of having dirs named based on single plugins -- in general (visual jailbreak refers also to the dataset name so no foul there)

I also dislike using the root for new artefacts directly (but this seems a pretty bland opinion :) )

I can also see that we have content that's used by plugins - currently all datasets - and content that's used by garak "core"

So we might like two top level dirs (not dead set on the names, not at all)

* `core` - things like plugin cache
* `datasets` - uh, datasets (`wn`, `visual_jailbreak`)

File: garak/resources/common.py
Comment (jmartin-tech) [reply garak/resources/common.py:55]: Thanks, due to this comment, I think I see a bug here as well since as coded `advbench_path` is only defined if the file existed in data already access to `advbench_path.is_file()` would raise exceptions.

At this time I am thinking of `data_path` as being `read_only` for runtime code, meaning anything downloaded is currently place in `cache_dir`.  This class currently exposes the one location where support for user override instead of download is being enabled and I am not sure yet if downloads should mix data in the a user provided override path.

I will leave the comment and fix the guard for this PR.

File: garak/probes/topic.py
Comment (leondz) [reply garak/probes/topic.py:95]: `nltk_data` i think might benefit from having its own handler; it's also used standalone. Maybe defer choice on this until we move the nltk, HF handlers out to their own place (e.g. `resources`)

File: garak/probes/topic.py
Comment (jmartin-tech) [reply garak/probes/topic.py:96]: > Where do we want to check for / make XDG dirs?
https://github.com/leondz/garak/blob/0e1d96e3892e14564aaece56c1ee9f92cfdd7dc0/garak/_config.py#L68-L71

`parents=True` is the equivalent to `-p`.

File: garak/resources/tap/tap_main.py
Comment (jmartin-tech) [reply garak/resources/tap/tap_main.py:38]: I agree these should be captured better for re-use. I will defer that change to a future iteration though. I think there is a larger issue to address as we have a few techniques that generate these intermediate files that could be valuable from a research context.

File: garak/payloads.py
Comment (jmartin-tech) [inline garak/payloads.py:56]: Utilizing the data_path object here will cause this to raise `GarakException` when the file is not found, as reflected in the test updates, and avoids the `FileNotFoundError` raised below.  Does it makes sense to allow the exception to bubble up or should the callers of `load_payload` be expected to handle the revised exception?

The error message on the exception will be `The resource requested does not exist {name}.json`.

I suspect the answer is that `Director.load()` should handle the `GarakException` with a log entry and raise `PayloadFailure` from that response.

Comment (jmartin-tech) [reply garak/payloads.py:54]: Updated in 958ea3c

File: garak/probes/topic.py
Comment (leondz) [reply garak/probes/topic.py:96]: Thanks, missed that. This looks good. Expect to move `wn` code to its own module/class once we know where those will live - did we settle on `resources`? Other options could be `apis`, `external`, `apps`, `common`, & _not_ `lib`. Or is this a conversation to conclude once this PR lands?

**Review by leondz — APPROVED** [review]
lgtm

File: garak/payloads.py
Comment (leondz) [reply garak/payloads.py:54]: Noting that the rglob in `payloads.Director` might conflict with other assumptions about payloads being file entries in `PAYLOAD_DIR`

**Comment by leondz** [comment]
NICE

**Pull request merged**
